interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,430 @@
+		return []byte(dbusConnectedPlugSecComp), nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		bus, name, err := iface.getAttribs(slot.Attrs)
+		if err != nil {
+			return nil, err
+		}
+
+		snippets := bytes.NewBufferString("")
+
+		// well-known DBus name-specific permanent slot policy
+		snippet := getAppArmorSnippet([]byte(dbusPermanentSlotAppArmor), bus, name)
+
+		// abstraction policy
+		abstraction, err := getAppArmorAbstraction(bus)
+		if err != nil {
+			return nil, err
+		}
+
+		old := []byte("###DBUS_ABSTRACTION###")
+		new := []byte(abstraction)
+		snippet = bytes.Replace(snippet, old, new, -1)
+
+		snippets.Write(snippet)
+
+		if release.OnClassic {
+			// classic-only policy
+			snippets.Write(getAppArmorSnippet([]byte(dbusPermanentSlotAppArmorClassic), bus, name))
+		}
+
+		return snippets.Bytes(), nil
+	case interfaces.SecuritySecComp:
+		return []byte(dbusPermanentSlotSecComp), nil
+	case interfaces.SecurityDBus:
+		bus, name, err := iface.getAttribs(slot.Attrs)
+		if err != nil {
+			return nil, err
+
+		}
+
+		// only system services need bus policy
+		if bus != "system" {
+			return nil, nil
+		}
+
+		old := []byte("###DBUS_NAME###")
+		new := []byte(name)
+		snippet := bytes.Replace([]byte(dbusPermanentSlotDBus), old, new, -1)
+
+		return []byte(snippet), nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	bus, name, err := iface.getAttribs(slot.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	busPlug, namePlug, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure that we only connect to slot with matching attributes
+	if bus != busPlug || name != namePlug {
+		return nil, nil
+	}
+
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// well-known DBus name-specific connected slot policy
+		snippet := getAppArmorSnippet([]byte(dbusConnectedSlotAppArmor), bus, name)
+
+		old := []byte("###PLUG_SECURITY_TAGS###")
+		new := plugAppLabelExpr(plug)
+		snippet = bytes.Replace(snippet, old, new, -1)
+
+		return snippet, nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	_, _, err := iface.getAttribs(plug.Attrs)
+	return err
+}
+
+func (iface *DbusInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	_, _, err := iface.getAttribs(slot.Attrs)
+	return err
+}
+
+func (iface *DbusInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) bool {
+	// allow what declarations allowed
+	return true
```

Comment:
the check // ensure that we only connect to slot with matching attributes
 +	if bus != busPlug || name != namePlug {

should probably be done here? instead?

that's what content is doing atm btw, until we don't have support for specifying attribute matching in the basedecl